tools/system_libs.py

```diff
@@ -93,6 +101,108 @@ def create_lib(libname, inputs):
 
 
+def create_ninja_file(input_files, filename, libname, cflags, asflags=None, customize_build_flags=None):
+  if asflags is None:
+    asflags = []
+  # TODO(sbc) There is an llvm bug that causes a crash when `-g` is used with
+  # assembly files that define wasm globals.
+  asflags = [arg for arg in asflags if arg != '-g']
+  cflags_asm = [arg for arg in cflags if arg != '-g']
+
+  def join(flags):
+    return ' '.join(flags)
+
+  out = f'''\
+# Automatically generated by tools/system_libs.py.  DO NOT EDIT
+
+ninja_required_version = 1.5
+
+ASFLAGS = {join(asflags)}
+CFLAGS = {join(cflags)}
+CFLAGS_ASM = {join(cflags_asm)}
+EMCC = {shared.EMCC}
+EMXX = {shared.EMXX}
+EMAR = {shared.EMAR}
+
+rule cc
+  depfile = $out.d
+  command = $EMCC -MD -MF $out.d $CFLAGS -c $in -o $out
+  description = CC $out
+
+rule cxx
+  depfile = $out.d
+  command = $EMXX -MD -MF $out.d $CFLAGS -c $in -o $out
+  description = CXX $out
+
+rule asm
+  command = $EMCC $ASFLAGS -c $in -o $out
+  description = ASM $out
+
+rule asm_cpp
+  depfile = $out.d
+  command = $EMCC -MD -MF $out.d $CFLAGS_ASM -c $in -o $out
+  description = ASM $out
+
+rule direct_cc
+  depfile = $with_depfile
+  command = $EMCC -MD -MF $with_depfile $CFLAGS -c $in -o $out
+  description = CC $out
+
+rule archive
+  command = $EMAR cr $out $in
+  description = AR $out
+
+'''
+  suffix = shared.suffix(libname)
+
+  if suffix == '.o':
+    assert len(input_files) == 1
+    depfile = shared.unsuffixed_basename(input_files[0]) + '.d'
+    out += f'build {libname}: direct_cc {input_files[0]}\n'
+    out += f'  with_depfile = {depfile}\n'
+  else:
+    case_insensitive = is_case_insensitive(build_dir)
+    objects = []
+    for src in input_files:
+      # Resolve duplicates by appending unique.
+      # This is needed on case insensitve filesystem to handle,
+      # for example, _exit.o and _Exit.o.
+      o = shared.unsuffixed_basename(src) + '.o'
+      object_uuid = 0
+      if case_insensitive:
+        o = o.lower()
+      # Find a unique basename
+      while o in objects:
+        object_uuid += 1
+        o = f'{o}__{object_uuid}.o'
+      objects.append(o)
+      ext = shared.suffix(src)
+      if ext == '.s':
+        out += f'build {o}: asm {src}\n'
+        flags = asflags
+      elif ext == '.S':
+        out += f'build {o}: asm_cpp {src}\n'
+        flags = cflags_asm
+      elif ext == '.c':
+        out += f'build {o}: cc {src}\n'
+        flags = cflags
+      else:
+        out += f'build {o}: cxx {src}\n'
+        flags = cflags
+      if customize_build_flags:
+        custom_flags = customize_build_flags(flags, src)
+        if custom_flags != flags:
+          out += f'  CFLAGS = {join(custom_flags)}'
+      out += '\n'
+
+    objects = sorted(objects, key=lambda x: os.path.basename(x))
+    objects = ' '.join(objects)
+    out += f'build {libname}: archive {objects}\n'
+
+  if not os.path.exists(filename) or utils.read_file(filename) != out:
```

Comment:
Is this really faster than just rewriting the ninja file unconditionally? Or is there some other reason it works this way?

Reply:
I was worried that re-writing the ninja file itself would cause stuff to get rebuilt.. but not that I tested again it seems that isn't necessary.  Reverting this part.